### DISCUSSION TITLE
[Markdown] Underscore for subscript does not work correctly when it is preceded by a curly brace

### DISCUSSION THREAD

**Discussion by TendTo** [community/community#65772] [Other Feature Feedback, Questions, & Ideas] [answered]
I had to write some math formulas, but I noticed some were not rendering as expected.
The culprit seems to be the `_` symbol. It is treated as the italic marker when it really should not.

The issue is solved by adding a `\` before `_`, ruining compatibility with other renderers.
When the `_` is preceded by any other letter or number, the issue does not arise.

#### Input
```tex
$$
\underbrace{192}_{8\ bit}.\underbrace{168}_{8\ bit}.\underbrace{1}_{8\ bit}.\underbrace{1}_{8\ bit}
$$
```

#### Expected output

The following Mathjax formula:

$`\underbrace{192}_{8\ bit}.\underbrace{168}_{8\ bit}.\underbrace{1}_{8\ bit}.\underbrace{1}_{8\ bit}`$

You can also verify by using another renderer like the one integrated into VsCode or copy and paste the formula into [this online tool](https://decadroid.github.io/mathjax-live-editor/) to see the expected results.

#### Real output
$$\underbrace{192}_{8\ bit}.\underbrace{168}_{8\ bit}.\underbrace{1}_{8\ bit}.\underbrace{1}_{8\ bit}$$

#### By @arafajs  

 Escape the underscores with backslashes.

##### Input

```tex
$\underbrace{192}\_{8\ bit}.\underbrace{168}\_{8\ bit}.\underbrace{1}\_{8\ bit}.\underbrace{1}\_{8\ bit}$
```

##### Output

$\underbrace{192}\_{8\ bit}.\underbrace{168}\_{8\ bit}.\underbrace{1}\_{8\ bit}.\underbrace{1}\_{8\ bit}$

##### Drawbacks

- It messes with other markdown renderers that do not have this problem to begin with

#### By @arafajs  

Using the math environment.

##### Input

<pre>
```math
\underbrace{192}_{8\ bit}.\underbrace{168}_{8\ bit}.\underbrace{1}_{8\ bit}.\underbrace{1}_{8\ bit}
```
</pre>

##### Output

```math
\underbrace{192}_{8\ bit}.\underbrace{168}_{8\ bit}.\underbrace{1}_{8\ bit}.\underbrace{1}_{8\ bit}
```

##### Drawbacks

- It messes with other markdown renderers that do not support this syntax

#### By @nschloe 

Add backticks to the formula.

```tex
$`...`$
```

##### Input

```tex
$`\underbrace{192}_{8\ bit}.\underbrace{168}_{8\ bit}.\underbrace{1}_{8\ bit}.\underbrace{1}_{8\ bit}`$
```

##### Output

$`\underbrace{192}_{8\ bit}.\underbrace{168}_{8\ bit}.\underbrace{1}_{8\ bit}.\underbrace{1}_{8\ bit}`$

##### Drawbacks

- It only works with the inline math notation (so no `$$...$$`)
- It messes with other markdown renderers that do not have this problem to begin with

**Comment by arafajs** [comment]
You're right, it seems like the underscores are being interpreted as italic markers when they should be subscript markers in this case.

- Escape the underscores with backslashes, as you mentioned. Not ideal, but it should render correctly:
```

$$
\underbrace{192}_{\textbackslash_8\ bit}. \underbrace{168}_{\textbackslash_8\ bit}. \underbrace{1}_{\textbackslash_8\ bit}. \underbrace{1}_{\textbackslash_8\ bit}  
$$
```

- Use a different subscript syntax like ^ or ~:

```

$$
\underbrace{192}^{8\ bit}.\underbrace{168}^{8\ bit}.\underbrace{1}^{8\ bit}.\underbrace{1}^{8\ bit}
```

- Put the subscripts in curly braces to group them:

```
$$
\underbrace{192}_{{8\ bit}}.\underbrace{168}_{{8\ bit}}.\underbrace{1}_{{8\ bit}}.\underbrace{1}_{{8\ bit}}
```

- Use \textsubscript from the amsmath package:

```
\usepackage{amsmath}
...
$$
\underbrace{192}_{\textsubscript{8\ bit}}.\underbrace{168}_{\textsubscript{8\ bit}}.
\underbrace{1}_{\textsubscript{8\ bit}}.\underbrace{1}_{\textsubscript{8\ bit}}  
$$
```

The root cause seems to be the parser getting confused by the curly brace before the underscore. I'd recommend reporting this as a bug/limitation to whoever created the rendering tools you're using.

**Reply by TendTo** [reply]
Thank you for the suggestions, but I needed this to work here on GitHub since I use that expression in a README.md file.
The same file renders perfectly when examined with VsCode or even when using other external MathJaX libraries but causes issues here on GitHub. If I were to fix it here with the first solution, it would break everywhere else.

**Reply by arafajs** [reply]
```
$$
\underbrace{192}_{{8 bit}}.\underbrace{168}_{{8 bit}}.\underbrace{1}_{{8 bit}}.\underbrace{1}_{{8 bit}} 
$$
```

**Reply by TendTo** [reply]
Does this work for you on the GitHub render? Because I tried your snippet, I still see it in italics.

**Reply by arafajs** [reply]
@TendTo GitHub Markdown parser does have some limitations compared to standard Markdown. There is no support for Markdown extras like table syntax, definition lists, footnote syntax, etc..

**Comment by nschloe** [comment]
This is a known bug, see, e.g., https://github.com/nschloe/github-math-bugs#inline-math-at-the-end-of-italic-text or https://github.com/nschloe/github-math-bugs#math-in-italic-text. The solution, as usual, is not to use the `$...$` syntax. Always use the backticked syntax
```markdown
$`...`$
```
This prevents the markdown santitation messing with the math content. Imho, it was a mistake on GitHub's side to try and support the dollar-only syntax in the first place.

**Reply by TendTo** [reply]
Thank you for the suggestion!
I did not know about that very long collection of GitHub math bugs; I will look there first next time.

The approach you proposed is a decent workaround, although it does not solve my problem:

- it does not seem to work in the multiline \$\$ ...  \$\$ environment
- it has the side effect of showing the backtick on other markdown renderers. Since they are my main target, I would like to avoid it

I'll make sure to update my question to include your suggestion.
I suppose I will have to wait until someone at GitHub fixes it.

**Reply by nschloe** [reply, marked as answer]
> it does not seem to work in the multiline $$ ... $$ environment

For multiline, use
````markdown
```math
a + b = c
```
````
Again, I think that the introduction `$$...$$`-syntax was a mistake.

> it has the side effect of showing the backtick on other markdown renderers. 

I'd call that a shortcoming of the other renderers. The backticked syntax is common over a number of platforms, e.g., GitLab. You might have better luck filing a bug at your other renderers.

**Reply by TendTo** [reply]
I like this approach. I can actually work with it. Thank you very much.
I have updated the question again.

**Reply by simonkeys** [reply]
> The solution, as usual, is not to use the ``$...$`` syntax. Always use the backticked syntax ``$`...`$``.

This doesn't seem reasonable, considering that GitHub's purpose is to host code that can be used on other platforms. Important systems like Pandoc, Quarto, Jupyterlab, etc., don't recognize the ``$`...`$`` syntax.

For the underscore issue, adding a regular space between close braces and underscores seems to solve the problem on GitHub without breaking rendering on other systems:

```
$$
\underbrace{192} _{8\ bit}.\underbrace{168} _{8\ bit}.\underbrace{1} _{8\ bit}.\underbrace{1} _{8\ bit}
$$
```
Output:

$$
\underbrace{192} _{8\ bit}.\underbrace{168} _{8\ bit}.\underbrace{1} _{8\ bit}.\underbrace{1} _{8\ bit}
$$
